### PR TITLE
SOLR-16824: Adopt Linux Command line tool pattern of -- for long option commands

### PR DESCRIPTION
**Pull request by epugh**
https://issues.apache.org/jira/browse/SOLR-16824

# Description

This moves us towards using `--help` style instead of the `-h` and `-help` mishmash that we have.  This also lines up with how Commons CLI wants to work as well.

# Solution

Update the various OptionBuilders, update the bats tests, and the Ref Guide.

# Tests

Running bats 

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by mkhludnev — APPROVED** [review]

**Comment by epugh** [comment]
@mkhludnev before I commit this to `main` only, are you able to give this a test on Windows?

File: solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc
Comment (janhoy) [inline solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc:51]: You can remove this line, as these changes are only from 10.0 onwards?

Comment (epugh) [reply solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc:51]: Correct...   This is way to hard to make be backward compatible...     Do I need to list out the before or after in the changelog, or would it be enough to say "See the Reference Guide page XXX for the long parameter names for each command."

**Comment by mkhludnev** [comment]
Hi, @epugh.
I started to explore this branch. Here we go: 
```
\dev-slim>bin\solr
...
Pass -help after any COMMAND to see command-specific usage information,
  such as:    solr start -help or solr stop -help
```

 I suppose we should have `--help` option here ^.

But the first option I tried, stumbled me. 

```
dev-slim>bin\solr start -h

ERROR: Hostname is required

Usage: solr start ....
```

I dont' think we have a chance to resolve the clash between host and help. 

```

\dev-slim>bin\solr start --help

Invalid command-line option: --help

Usage: solr start ...
```

```
dev-slim>bin\solr start -help

Usage: solr start [-f] ...
```

I seems like one dash is working.

**Comment by epugh** [comment]
I think this branch has definitly had some bit rot, and there is a lot of work to update it....  Thanks for looking @mkhludnev at this branch, glad to get some input on this!

**Comment by epugh** [comment]
I updated this PR (thanks StaleBot for bugging me) from Main, and I think have to redo some changes...

**Comment by epugh** [comment]
Before calling this done, make sure to check out https://github.com/apache/solr/pull/1768/commits/395eaab7199f5f531f6a4121f4e946456ff5c5ad and make sure all the reverted changes (when bringing this PR up to date) are reapplied!

File: solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc
Comment (epugh) [reply solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc]: I've made the change that @janhoy  suggested!

**Comment by epugh** [comment]
Pending a good run off all the tests, I think I'm done and would love some review.   This is to go into Solr X and not be backported to Solr 9...

File: solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc
Comment (janhoy) [inline solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc:31]:
```suggestion
The Solr CLI has gone through some significant renovations to reduce technical debt, and now functions more consistently and predictably. Most notably, CLI commands now use unix-style options, e.g. `--help` instead of `-help`.
```

File: solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc
Comment (janhoy) [inline solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-10.adoc:32]:
```suggestion
Users are urged to review all use of the `bin/solr` command, such as in scripts and documentation, to ensure that the correct options are being used.
```

**Review by janhoy — COMMENTED** [review]
Too large to review it all in detail.

I updated the refguide upgrade notes a bit.

Ideally there would be a deprecation before removal of old opts, but I realize how that would blow up this PR with not only a bunch of extra Options but also logic to parse old and new and print deprecation warnings etc. It's a pity commons-cli does not support a `deprecatedShortOpt()` and `deprecatedLongOpt()` settings that would handle such logic automatically.

**Comment by epugh** [comment]
Thanks @janhoy for starting to look.  And yeah, this is 10x only change.  Trying to manage two sets of old and new parameters I think would just make this impossible ;-).  I'll leave this open for a few more days to give folks time to look at it more.

File: solr/bin/solr.cmd
Comment (madrob) [inline solr/bin/solr.cmd:553]: double-dash?

**Review by madrob — COMMENTED** [review]
Is it possible to back port the non-conflicting ones to 9.x? And also log deprecation warnings on anything that's going away?

**Comment by epugh** [comment]
> Is it possible to back port the non-conflicting ones to 9.x? And also log deprecation warnings on anything that's going awa

**Comment by epugh** [comment]
> Is it possible to back port the non-conflicting ones to 9.x? And also log deprecation warnings on anything that's going away?

I was sort of hoping not too...    I *guess* that I could backport places where we ADD something, like a new `--long-option`.  My thought was to document in the upgrade notes more about the changes, but not try to navigate the changes to the shell and command scirpts and the java code...    

What's your temperature on this?  Worth discussing at the community meetup and getting more input?

**Comment by madrob** [comment]
My ideal world is everywhere that we remained -x to -y we add -y to 9.x and log a warning on usage of -x. Similar to what gradle does.

But I can appreciate that that’s a ton of work, so maybe we can do that with just the common ones? zkHost at least?

**Comment by epugh** [comment]
> My ideal world is everywhere that we remained -x to -y we add -y to 9.x and log a warning on usage of -x. Similar to what gradle does.
> 
> But I can appreciate that that’s a ton of work, so maybe we can do that with just the common ones? zkHost at least?

I will reach out to the commons-cli community see if there are any suggestions.   I don't know if we can `-z` and `-zkHost` as short parameters for the same option.  And if we just start logging in 9.x "Heads up -zkHost will be --zk-host in 10x, but you can't use it today", then that might be annonying.

Once on 10, if you attempt to use `-zkHost` from muscle memory, you will get an error immediately until you swap to `-z` or `--zk-host`.

**Comment by janhoy** [comment]
I have been thinking about adding a deprecation-feature to commons-cli, such that you could construct an Option: 
```java
Option o = Option.builder("n").longOpt("new-long").deprecatedOpt("oldShort").deprecatedLongOpt("oldLong")
...
String myVal = cli.getOptionValue("n")
```

It would work so that if `-n` or `--new-long` are not provided, it will fall back to checking `-oldShort` and `--oldLong`, and print a deprecation warning if using deprecated options. Perhaps we could help implement it and use the new version before the 10.0 release..

**Comment by epugh** [comment]
One reason I've been feeling a bit daunted on the idea of enhancing commons-cli to support deprecation is because it would mean that we would need to keep/extend the lines of code in `bin/solr` and `bin/solr.cmd` that deal with parameter parsing...  Although now that I look at it, maybe if we move along more of the parameter passing INTO the java code, then the impact would be less.

**Comment by epugh** [comment]
I have something cooking for c

> I have been thinking about adding a deprecation-feature to commons-cli, such that you could construct an Option:
> 
> ```java
> Option o = Option.builder("n").longOpt("new-long").deprecatedOpt("oldShort").deprecatedLongOpt("oldLong")
> ...
> String myVal = cli.getOptionValue("n")
> ```
> 
> It would work so that if `-n` or `--new-long` are not provided, it will fall back to checking `-oldShort` and `--oldLong`, and print a deprecation warning if using deprecated options. Perhaps we could help implement it and use the new version before the 10.0 release..

I have something cooking along these lines!  It actually helped me find places where I was using `solrUrl` instead of `solr-url` as the parameter name in the CLI!

**Comment by epugh** [comment]
> > I have been thinking about adding a deprecation-feature to commons-cli, such that you could construct an Option:
> > ```java
> > Option o = Option.builder("n").longOpt("new-long").deprecatedOpt("oldShort").deprecatedLongOpt("oldLong")
> > ...
> > String myVal = cli.getOptionValue("n")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It would work so that if `-n` or `--new-long` are not provided, it will fall back to checking `-oldShort` and `--oldLong`, and print a deprecation warning if using deprecated options. Perhaps we could help implement it and use the new version before the 10.0 release..
> 
> I have something cooking along these lines! It actually helped me find places where I was using `solrUrl` instead of `solr-url` as the parameter name in the CLI!

https://issues.apache.org/jira/browse/CLI-329

**Comment by epugh** [comment]
What do folks think about moving forward with this as is, and then adding the deprecations when commons-cli comes out....

**Comment by epugh** [comment]
https://github.com/apache/commons-cli/pull/252 replaced what I proposed and was merged.  Now waiting on a release!

**Comment by epugh** [comment]
Waiting for the upcoming commons cli 1.7.1 to come out with nicer handling of deprecated options.

**Comment by janhoy** [comment]
1.8.0 is out, will merge the renovate PR https://github.com/apache/solr/pull/2481

**Review by janhoy — COMMENTED** [review]
I cherry-picked the commons-cli 1.8.0 commit into this PR (since it was too far off from main for a quick merge).

This approach with deprecated options looks promising!

**Comment by epugh** [comment]
@janhoy  I ran a bit out of steam while waiting for 1.8.0, please do feel free to continue to push this PR along ;-).

**Comment by janhoy** [comment]
I had a thought now that `bin/solr` supports a remote Solr cluster with `--solr-url`. We currently support `solr version` which prints the version from `SolrVersion.java`, i.e. the client version. However, if you run the tool towards a remote solr with a different version, you won't see that.

So I wonder if `solr --version` should be the tool version, but `solr version --solr-url http://foo.bar/solr/` should call the backend to ask for its version.

Probably scope creep for this PR though.

**Comment by epugh** [comment]
> I had a thought now that `bin/solr` supports a remote Solr cluster with `--solr-url`. We currently support `solr version` which prints the version from `SolrVersion.java`, i.e. the client version. However, if you run the tool towards a remote solr with a different version, you won't see that.
> 
> So I wonder if `solr --version` should be the tool version, but `solr version --solr-url http://foo.bar/solr/` should call the backend to ask for its version.
> 
> Probably scope creep for this PR though.

i love it...    https://issues.apache.org/jira/browse/SOLR-17318

**Comment by janhoy** [comment]
Did a few low hanging ones

**Comment by epugh** [comment]
Hey all..   so, just got all green on the checks!   I'm wondering, how many of the parameters do we want to have deprecation for...  We have it for solr url and zk host, and they are pretty common.  Do we want to add it for all the other parameters?   I'm on the fence about it, because each one requires additional code, and they are each relatively hard to test...

**Comment by epugh** [comment]
chatted with @gerlowskija  and the plan is to mark all changed features as Deprecated, not just the solrurl and zkhost, for 9.x, so your scripts don't break in 9.6 to 9.7, but to remove the deprecated option support in 10x....

**Comment by epugh** [comment]
Okay!  I believe that deprecated options are back in, so scripts shouldn't break between 9.6 and 9.7....   I would love another set of eyes and then look to merge to main in the next few days.

**Comment by epugh** [comment]
@HoustonPutman any tips on how to validate that changes on this PR don't impact solr operator?

File: solr/core/src/test/org/apache/solr/cli/AuthToolTest.java
Comment (gerlowskija) [inline solr/core/src/test/org/apache/solr/cli/AuthToolTest.java:67]: [0] You've put in a lot of effort in this PR to make sure that we (at least temporarily) maintain backcompat with older options.  Given all that work, it seems a shame to not validate it.

One way we could do that is by randomizing which form of different options gets used in these tests.  Some of the more common CLI parameters could be replaced by calls to a method like:

```
public void getZooKeeperParameter() {
    if (rarely()) return "-z";
    if (rarely()) return "-zkhost";
    return "--zk-host";
}
```

Not important enough to require on this PR, but worth considering if you like the idea.

**Review by gerlowskija — COMMENTED** [review]
Wow - this looks like a real Herculean effort @epugh - thanks for working on this.

I'm about 2/3 through the PR, but I think I've left the bulk of my comments.  Mostly minor things: places where we should probably be kebab-casing for consistency, etc.  Will finish my review through this morning, but wanted to get the bulk of my comments out there for quicker iteration -- I know you're looking to get this in.

File: solr/solr-ref-guide/modules/deployment-guide/pages/solr-control-script-reference.adoc
Comment (gerlowskija) [inline solr/solr-ref-guide/modules/deployment-guide/pages/solr-control-script-reference.adoc:854]: [0] Missing some double-dashes here.

File: solr/solr-ref-guide/modules/deployment-guide/pages/solr-control-script-reference.adoc
Comment (gerlowskija) [inline solr/solr-ref-guide/modules/deployment-guide/pages/solr-control-script-reference.adoc:1834]: [0] Double-dash

File: solr/test-framework/src/java/org/apache/solr/cloud/AbstractDistribZkTestBase.java
Comment (gerlowskija) [inline solr/test-framework/src/java/org/apache/solr/cloud/AbstractDistribZkTestBase.java:418]: [0] Double dashes for confname and confdir?

**Review by gerlowskija — COMMENTED** [review]
Some additional comments

**Comment by epugh** [comment]
@janhoy  If you have some bandwidth, I'm a bit time constrained today and tomorrow...    And would love to have it ready to merge by Wednesday EST so that it has a day or two for any issues that come up before I head out on vacation.   I saw that there are some merge conflicts ;-(

File: solr/bin/solr
Comment (gerlowskija) [reply solr/bin/solr:1422]: I talked to @epugh offline to get a better understanding of this.

My main takeaway from that discussion was that we should probably accept the single-dash *and* double-dash variant in most of these places.  Some locations don't require explicit single-dash handling, because of how the option is passed through to the Java code and handled there.  But in many cases we do, so it's probably best to handle explicit single-dash handling in most instances.

(Correct me if that's wrong @epugh )

File: solr/core/src/test/org/apache/solr/cli/AuthToolTest.java
Comment (gerlowskija) [reply solr/core/src/test/org/apache/solr/cli/AuthToolTest.java:67]: See SOLR-17344 - we can discuss this issue further there.

File: solr/bin/solr
Comment (gerlowskija) [reply solr/bin/solr:440]: In hindsight this seems low-priority IMO.  Other CLI tools aren't necessarily consistent in what opt-variants they offer from param to param.  But I created a JIRA ticket in case it strikes someone as important - SOLR-17345.

File: solr/core/src/java/org/apache/solr/cli/SolrCLI.java
Comment (epugh) [reply solr/core/src/java/org/apache/solr/cli/SolrCLI.java:147]: Thanks!

**Comment by epugh** [comment]
@gerlowskija I finished kebabing ;-).   I won't say I got every example everywhere, but any more could be a follow on!

**Comment by epugh** [comment]
For CHANGES.txt:

```
Adopt Linux standard pattern of -- for long option commands that are kebab based.  I.e -zkHost is now --zk-host.
```

**Pull request merged**
